README.md

![scroll-into-view](/scrollintoview.png) ![example-gif](/scrollIntoViewExample.gif)

[![Build Status](https://travis-ci.org/KoryNunn/scroll-into-view.svg?branch=master)](https://travis-ci.org/KoryNunn/scroll-into-view)
[![Backers on Open Collective](https://opencollective.com/scroll-into-view/backers/badge.svg)](#backers)
[![Sponsors on Open Collective](https://opencollective.com/scroll-into-view/sponsors/badge.svg)](#sponsors)
## What

Scrolls an element into view

Also scrolls any scroll-parents so that the element is in view.

## Donating

If you want to show your support financially, [I'm on Patreon](https://patreon.com/user/korynunn)

## Example

[Example](http://korynunn.github.io/scroll-into-view/example/)

## How

require it
```javascript
var scrollIntoView = require('scroll-into-view');
```
use it

```javascript
scrollIntoView(someElement);
```

You can pass settings to control the time, easing, and whether or not a parent is a valid element to scroll, and alignment:

All options are optional.

```javascript
scrollIntoView(someElement, {

    time: 500, // half a second


    ease: function(value){
        return Math.pow(value,2) - value; // Do something weird.
    },

    validTarget: function(target, parentsScrolled){

        // Only scroll the first two elements that don't have the class "dontScroll"
        // Element.matches is not supported in IE11, consider using Element.prototype.msMatchesSelector if you need to support that browser

        return parentsScrolled < 2 && target !== window && !target.matches('.dontScroll');
    },

    align:{
        top: 0 to 1, default 0.5 (center)
        left: 0 to 1, default 0.5 (center)
        topOffset: pixels to offset top alignment
        leftOffset: pixels to offset left alignment
    },

    isScrollable: function(target, defaultIsScrollable){

        // By default scroll-into-view will only attempt to scroll elements that have overflow not set to `"hidden"` and who's scroll width/height is larger than their client height.
        // You can override this check by passing an `isScrollable` function to settings:

        return defaultIsScrollable(target) || ~target.className.indexOf('scrollable');
    },

    isWindow: function(target){
        // If you need special detection of the window object for some reason, you can do it here.
        return target.self === target;
    },

    cancellable: true, // default is true, set to false to prevent users from cancelling the scroll with a touch or mousewheel

    maxSynchronousAlignments: 3, // default is 3. Maximum number of times to try and align elements synchronously before completing.

    debug: true // default is false. This will spit out some logs that can help you understand what scroll-into-view is doing if it isn't doing what you expect.
});
```

You can pass a callback that will be called when all scrolling has been completed or canceled.

```javascript
scrollIntoView(someElement [, settings], function(type){
    // Scrolling done.
    // type will be 'complete' if the scroll completed or 'canceled' if the current scroll was canceled by a new scroll
});
```

## Size

Small. ~3.03 KB for the standalone.

## Changelog

[View Changeog](https://changelogit.korynunn.com/#korynunn/scroll-into-view)

## Testing

Testing scrolling is really hard without stuff like webdriver, but what's there works ok for the moment.

The tests will attempt to launch google-chrome. If you don't have chrome, lol just kidding you do.

```
npm run test
```

# Standalone

If you want to use this module without browserify, you can use `scrollIntoView.min.js`

```
<script src="scrollIntoView.min.js"></script>

<script>
    window.scrollIntoView(someElement);
</script>
```

## Contributors

This project exists thanks to all the people who contribute.
<a href="https://github.com/KoryNunn/scroll-into-view/graphs/contributors"><img src="https://opencollective.com/scroll-into-view/contributors.svg?width=890&button=false" /></a>


## Backers

Thank you to all our backers! 🙏 [[Become a backer](https://opencollective.com/scroll-into-view#backer)]

<a href="https://opencollective.com/scroll-into-view#backers" target="_blank"><img src="https://opencollective.com/scroll-into-view/backers.svg?width=890"></a>


## Sponsors

Support this project by becoming a sponsor. Your logo will show up here with a link to your website. [[Become a sponsor](https://opencollective.com/scroll-into-view#sponsor)]

<a href="https://opencollective.com/scroll-into-view/sponsor/0/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/0/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/1/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/1/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/2/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/2/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/3/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/3/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/4/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/4/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/5/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/5/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/6/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/6/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/7/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/7/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/8/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/8/avatar.svg"></a>
<a href="https://opencollective.com/scroll-into-view/sponsor/9/website" target="_blank"><img src="https://opencollective.com/scroll-into-view/sponsor/9/avatar.svg"></a>




CHANGELOG.md

## 1.12.3 - Thu Nov 21 2019

#73 Improved scrolling in direct children of body - [1be82cb8](João Ramos@1be82cb887e1363a906925084e4b3e2183ecc133)


## 1.12.2 - Thu Oct 24 2019

Testing ci setup - [dab9b219](korynunn@dab9b219f6ade57adf57f3b5dafdec0b5ab2b410)

Update ci file for headless testing - [3219b9bf](korynunn@3219b9bf34c2e9242f5c57c1136a524d477d50f5)

Add status badge - [9e6062a6](korynunn@9e6062a6ca7770ae9a5673f79660d291dc5dc111)

Fix readme - [ee627fde](korynunn@ee627fde69a4eda1b4e9c1a712809f35e3e9c505)

Moved badge again - [e735fe1e](korynunn@e735fe1e403eff68915b20611322f1411f1f2b56)

Updated changelog - [b037ed6d](korynunn@b037ed6dbc31247c9f43a847612d66574bcbcdbf)


## 1.12.1 - Thu Oct 24 2019

Update package description - [ca09d990](korynunn@ca09d990047bdb796ba7dc4629da715f3b215ed1)


## 1.12.0 - Thu Oct 24 2019

Fixed potential overflow bug, added maxSynchronousAlignments, cleaned package - [566a2480](korynunn@566a2480d7f707bea06c0cf92052b5e2cb3243f6)


## 1.11.0 - Fri Oct 11 2019

Add cancellable option - [13ed1cb7](korynunn@13ed1cb79436908f68ba566be9c2ecc8f2e5f3e3)


## 1.10.0 - Fri Oct 11 2019

Added isWindow option - [932a23a0](korynunn@932a23a0f0a4e5f8fcb6c74c9c4ff5ad726e765b)


## 1.9.7 - Tue May 7 2019

Removed opencollective postinstall - [8b63c9ad](korynunn@8b63c9adb6de93dde7157ef0bd7e2f205873fb43)


## 1.9.6 - Fri Apr 19 2019

Add files via upload - [650af702](Kory Nunn@650af7020db3a72198f2fc5c4efa823ab701a337)

Fire 'complete' even if no elements are scrolled - [902069a2](Kory Nunn@902069a2fd1306f425b520213578c4019da31294)


## 1.9.5 - Fri Mar 22 2019

Add changelog on precommit - [26489fc2](korynunn@26489fc22c58988240567213e010803a107e3666)


## 1.9.4 - Thu Mar 14 2019

Added backers and sponsors on the README - [8bb98847](Jess@8bb98847f59b3c37ed479adbab7cac2c6800e56f)

Added call to donate after npm install (optional) - [6d6b4805](Jess@6d6b4805f5fb14e9055a06087d63fa1579f6b220)

Manual scroll should cancel/override scrollIntoView. Fixes #58. - [d9a86bac](Dane David@d9a86baca2c4bc87c3e6e430495c99b150802b31)


## 1.9.3 - Fri Mar 23 2018

Fixes #10 - [712cd284](Kory Nunn@712cd28432dd634e2899719e929ec3fbcd0fb350)


## 1.9.2 - Fri Mar 23 2018

Update package.json license to be consistent with license in project. Add note within readme about IE11 compatability with method used - [b027edfa](Marcus Wood@b027edfafa183117a8a2ef8bbbf566a0b70440e9)

update readme - [a710fc2a](mwood23@a710fc2ab8c6253b579018ade3b9625c4677c3e6)

Change to ms prefixed method - [369a510a](mwood23@369a510abade9059b3027222e1b84b541a3e98f3)

Added tests for 0 time (instant) scroll. - [34c32a40](Kory Nunn@34c32a40df77e8f50989e1b3f4cef423327746d2)

Added passive: true to handlers - [772d67b4](Kory Nunn@772d67b4f31bb2d1d2c5d94548bc64b03afa3633)


## 1.9.1 - Mon Sep 18 2017

Added logo - [dd6b9708](Kory Nunn@dd6b970852c321ed5bc39c6caa28ccd55827f64b)

Correct logo - [2db2f10e](Kory Nunn@2db2f10ec83186e54a658f3b7fcf7d50d151fbef)

feat: Add iframe support - [f667da54](Craig Michael Thompson@f667da5456f8c1f58335a15aa13f700fd87bbbe5)

Change to .self check instead of pageXOffset - [307f257e](Craig Michael Thompson@307f257e23cd93541e1e2011278dfa51202b0a89)

Fixed iframe usage - [29f4f9c5](Kory Nunn@29f4f9c5d41a5e33a921259cf7862b9bc7780e55)


## 1.9.0 - Wed Jul 19 2017

settings.validScrollable - [8519abbd](Vladimir Barbarosh@8519abbdbacd5e2297ef66cb0b79180b073dccdc)

add test for validScrollable - [0bf72315](Vladimir Barbarosh@0bf72315c349b5a7592a2bd0e5adaa6d1230a20b)

move check for firefox outside of queue fn - [5ab28568](Vladimir Barbarosh@5ab285688752389d90799090d1fa6e8bf352af40)

Pulled in vbarbarosh's changes, added isScrollable api - [2cd148fc](Kory Nunn@2cd148fce31020c00a6216010c0e7bbb366bdade)


## 1.8.3 - Wed Jul 12 2017

Fixed absolute positioning bug - [4ab91e7b](Kory Nunn@4ab91e7b5c4ec4469fad4a53d2f4f75304f3ff67)


## 1.8.2 - Thu Jun 22 2017

Added built version - [4c78837d](Kory Nunn@4c78837d2d1b2031fd602bbc765b4250738616c8)


## 1.8.1 - Thu Jun 22 2017

Fixed wrong variable name issue - [db153b50](Kory Nunn@db153b509cefbcd95d587e18287ead84f71c15ec)


## 1.8.0 - Mon Apr 3 2017

Set fixed header height - [5ac918bb](Eric van Riet Paap@5ac918bb2b11b15da9614d5059202f6ba80be38f)

align.leftOffset & align.topOffset can by used - [94d53223](Eric van Riet Paap@94d53223ca3b02bf62a92c19f310c5010a84ad80)

Normalize offsets like align.top and friends - [998bb172](Eric van Riet Paap@998bb172330df69eb20d6192c0988586b2e3df53)

Document topOffset and leftOffset - [2b134148](Eric van Riet Paap@2b134148f6ae28463dc39b9d47616a5672da0012)

made pixel offset - [628a4b82](Kory Nunn@628a4b822749a5b23813def58e77e9aec2ef2e74)


## 1.7.6 - Thu Mar 23 2017

Remove erroneous apostrophe from README - [096ec4b4](ray hatfield@096ec4b47df4706383fd0c3d4714cc68defd67d1)

Fixed logical check for validTarget - [956ee6f7](Donovan Isherwood@956ee6f7eb227184eb1c2dea11677e01a00fdab5)

Cleanup - [b25c2ca8](Kory Nunn@b25c2ca8133c957d869751d34f1da4a1358dc041)


## 1.7.5 - Mon Feb 20 2017

Use window.pageYOffset over window.scrollY has it is cross browser - [32d17d59](Nicolas Duvieusart Déry@32d17d590ef5357bfe8da85ed0db87eebd9d6d9e)

Updated built files - [db06d859](Kory Nunn@db06d859fafe48339e2a45d32acb88a7b7337ea3)


## 1.7.4 - Wed Feb 8 2017

generified the readme even more - [b496e308](Kory Nunn@b496e308cecb04d141eb9bdafe3151152e77bd82)


## 1.7.3 - Mon Feb 6 2017

Removed log - [1ceac5fd](Kory Nunn@1ceac5fd183105ffdbc4e03e4db24cb0a43b4343)


## 1.7.2 - Mon Feb 6 2017

Readability - [7360f656](Kory Nunn@7360f656ae6ed38f618a89002893aeff83b5c59b)

Updated readme, fixed easing kinda, resolves #12 - [6c6ac24b](Kory Nunn@6c6ac24ba3169bc55307da72e6f597913bb345fc)

fixed example - [bd044d80](Kory Nunn@bd044d80f0766ac3a96026920ece3c73d10dc946)

Updated example link - [05f43730](Kory Nunn@05f4373095d643bab4b5908c8de27f30d5d0aba7)

Fixed leaking touchstart handler - [5095dffd](Kory Nunn@5095dffd8aeda2073714faa48b27c844c8418750)



test/test2.js
window.scrollIntoView = require('../');

example/index.js
var crel = require('crel'),
    scrollIntoView = require('../');

window.addEventListener('load', function(){
    var menu;
    crel(document.body,
        crel('div', {'style':'position:absolute; height:300%; width: 100%'},
            crel('div', {'style':'position:absolute; width:300%; height: 300px; top:50%;'},
                crel('div', {'style':'position:absolute; height:200%; width: 300px; left:50%'},
                    target = crel('span', {'style':'position:absolute; top:50%; width: 150px'})
                )
            )
        ),
        menu = crel('nav', { style: 'position:fixed; top:0; left:0; bottom: 0; width: 200px; border: solid 4px white;' })
    );

    function align(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true, time: 1000}, function(type){
            target.textContent = type;
        });
    }
    align();

    function uncancellableAlign(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true,  time: 2000, cancellable: false }, function(type){
            target.textContent = type;
        });
    }

    function ease(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true,
            time: 1000,
            align:{
                top: 0,
                left: 0.5
            },
            ease: function(value){
                return 1 - Math.pow(1 - value, value / 5);
            }
        }, function(type){
            target.textContent = type;
        });
    }

    var side = 1;
    function buttonText(){
        return 'align ' + (side ? 'right' : 'left') + ' edge with menu';
    }
    function menuAlign(){
        target.textContent = 'scrolling';
        scrollIntoView(target, {
            debug: true,
            time: 1000,
            align:{
                top: 0.5,
                left: 0,
                leftOffset: side ? 200 : 40,
                topOffset: 0
            }
        }, function(type){
            target.textContent = type;
        });
        side = (side + 1) % 2;
        menuAlignButton.textContent = buttonText();
    }

    var alignButton,
        uncancellableAlignButton,
        easeButton,
        menuAlignButton;

    crel(menu,
        alignButton = crel('button', {'style':'width: 190px'},
            'scroll into view'
        ),
        uncancellableAlignButton = crel('button', {'style':'width: 190px'},
            'uncancellable scroll into view'
        ),
        easeButton = crel('button', {'style':'width: 190px'},
            'scroll into view with custom easing'
        ),
        menuAlignButton = crel('button', {'style':'width: 190px'},
            buttonText()
        )
    );

    alignButton.addEventListener('click', align);
    uncancellableAlignButton.addEventListener('click', uncancellableAlign);
    easeButton.addEventListener('click', ease);
    menuAlignButton.addEventListener('click', menuAlign);
});

example/index.browser.js
(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
var crel = require('crel'),
    scrollIntoView = require('../');

window.addEventListener('load', function(){
    var menu;
    crel(document.body,
        crel('div', {'style':'position:absolute; height:300%; width: 100%'},
            crel('div', {'style':'position:absolute; width:300%; height: 300px; top:50%;'},
                crel('div', {'style':'position:absolute; height:200%; width: 300px; left:50%'},
                    target = crel('span', {'style':'position:absolute; top:50%; width: 150px'})
                )
            )
        ),
        menu = crel('nav', { style: 'position:fixed; top:0; left:0; bottom: 0; width: 200px; border: solid 4px white;' })
    );

    function align(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true, time: 1000}, function(type){
            target.textContent = type;
        });
    }
    align();

    function uncancellableAlign(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true,  time: 2000, cancellable: false }, function(type){
            target.textContent = type;
        });
    }

    function ease(){
        target.textContent = 'scrolling';
        scrollIntoView(target, { debug: true,
            time: 1000,
            align:{
                top: 0,
                left: 0.5
            },
            ease: function(value){
                return 1 - Math.pow(1 - value, value / 5);
            }
        }, function(type){
            target.textContent = type;
        });
    }

    var side = 1;
    function buttonText(){
        return 'align ' + (side ? 'right' : 'left') + ' edge with menu';
    }
    function menuAlign(){
        target.textContent = 'scrolling';
        scrollIntoView(target, {
            debug: true,
            time: 1000,
            align:{
                top: 0.5,
                left: 0,
                leftOffset: side ? 200 : 40,
                topOffset: 0
            }
        }, function(type){
            target.textContent = type;
        });
        side = (side + 1) % 2;
        menuAlignButton.textContent = buttonText();
    }

    var alignButton,
        uncancellableAlignButton,
        easeButton,
        menuAlignButton;

    crel(menu,
        alignButton = crel('button', {'style':'width: 190px'},
            'scroll into view'
        ),
        uncancellableAlignButton = crel('button', {'style':'width: 190px'},
            'uncancellable scroll into view'
        ),
        easeButton = crel('button', {'style':'width: 190px'},
            'scroll into view with custom easing'
        ),
        menuAlignButton = crel('button', {'style':'width: 190px'},
            buttonText()
        )
    );

    alignButton.addEventListener('click', align);
    uncancellableAlignButton.addEventListener('click', uncancellableAlign);
    easeButton.addEventListener('click', ease);
    menuAlignButton.addEventListener('click', menuAlign);
});
},{"../":3,"crel":2}],2:[function(require,module,exports){
//Copyright (C) 2012 Kory Nunn

//Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

/*

    This code is not formatted for readability, but rather run-speed and to assist compilers.

    However, the code's intention should be transparent.

    *** IE SUPPORT ***

    If you require this library to work in IE7, add the following after declaring crel.

    var testDiv = document.createElement('div'),
        testLabel = document.createElement('label');

    testDiv.setAttribute('class', 'a');
    testDiv['className'] !== 'a' ? crel.attrMap['class'] = 'className':undefined;
    testDiv.setAttribute('name','a');
    testDiv['name'] !== 'a' ? crel.attrMap['name'] = function(element, value){
        element.id = value;
    }:undefined;


    testLabel.setAttribute('for', 'a');
    testLabel['htmlFor'] !== 'a' ? crel.attrMap['for'] = 'htmlFor':undefined;



*/

(function (root, factory) {
    if (typeof exports === 'object') {
        module.exports = factory();
    } else if (typeof define === 'function' && define.amd) {
        define(factory);
    } else {
        root.crel = factory();
    }
}(this, function () {
    // based on http://stackoverflow.com/questions/384286/javascript-isdom-how-do-you-check-if-a-javascript-object-is-a-dom-object
    var isNode = typeof Node === 'function'
        ? function (object) { return object instanceof Node; }
        : function (object) {
            return object
                && typeof object === 'object'
                && typeof object.nodeType === 'number'
                && typeof object.nodeName === 'string';
        };
    var isArray = function(a){ return a instanceof Array; };
    var appendChild = function(element, child) {
      if(!isNode(child)){
          child = document.createTextNode(child);
      }
      element.appendChild(child);
    };


    function crel(){
        var document = window.document,
            args = arguments, //Note: assigned to a variable to assist compilers. Saves about 40 bytes in closure compiler. Has negligable effect on performance.
            element = args[0],
            child,
            settings = args[1],
            childIndex = 2,
            argumentsLength = args.length,
            attributeMap = crel.attrMap;

        element = isNode(element) ? element : document.createElement(element);
        // shortcut
        if(argumentsLength === 1){
            return element;
        }

        if(typeof settings !== 'object' || isNode(settings) || isArray(settings)) {
            --childIndex;
            settings = null;
        }

        // shortcut if there is only one child that is a string
        if((argumentsLength - childIndex) === 1 && typeof args[childIndex] === 'string' && element.textContent !== undefined){
            element.textContent = args[childIndex];
        }else{
            for(; childIndex < argumentsLength; ++childIndex){
                child = args[childIndex];

                if(child == null){
                    continue;
                }

                if (isArray(child)) {
                  for (var i=0; i < child.length; ++i) {
                    appendChild(element, child[i]);
                  }
                } else {
                  appendChild(element, child);
                }
            }
        }

        for(var key in settings){
            if(!attributeMap[key]){
                element.setAttribute(key, settings[key]);
            }else{
                var attr = crel.attrMap[key];
                if(typeof attr === 'function'){
                    attr(element, settings[key]);
                }else{
                    element.setAttribute(attr, settings[key]);
                }
            }
        }

        return element;
    }

    // Used for mapping one kind of attribute to the supported version of that in bad browsers.
    // String referenced so that compilers maintain the property name.
    crel['attrMap'] = {};

    // String referenced so that compilers maintain the property name.
    crel["isNode"] = isNode;

    return crel;
}));

},{}],3:[function(require,module,exports){
var COMPLETE = 'complete',
    CANCELED = 'canceled';

function raf(task){
    if('requestAnimationFrame' in window){
        return window.requestAnimationFrame(task);
    }

    setTimeout(task, 16);
}

function setElementScroll(element, x, y){
    if(element.self === element){
        element.scrollTo(x, y);
    }else{
        element.scrollLeft = x;
        element.scrollTop = y;
    }
}

function getTargetScrollLocation(scrollSettings, parent){
    var align = scrollSettings.align,
        target = scrollSettings.target,
        targetPosition = target.getBoundingClientRect(),
        parentPosition,
        x,
        y,
        differenceX,
        differenceY,
        targetWidth,
        targetHeight,
        leftAlign = align && align.left != null ? align.left : 0.5,
        topAlign = align && align.top != null ? align.top : 0.5,
        leftOffset = align && align.leftOffset != null ? align.leftOffset : 0,
        topOffset = align && align.topOffset != null ? align.topOffset : 0,
        leftScalar = leftAlign,
        topScalar = topAlign;

    if(scrollSettings.isWindow(parent)){
        targetWidth = Math.min(targetPosition.width, parent.innerWidth);
        targetHeight = Math.min(targetPosition.height, parent.innerHeight);
        x = targetPosition.left + parent.pageXOffset - parent.innerWidth * leftScalar + targetWidth * leftScalar;
        y = targetPosition.top + parent.pageYOffset - parent.innerHeight * topScalar + targetHeight * topScalar;
        x -= leftOffset;
        y -= topOffset;
        differenceX = x - parent.pageXOffset;
        differenceY = y - parent.pageYOffset;
    }else{
        targetWidth = targetPosition.width;
        targetHeight = targetPosition.height;
        parentPosition = parent.getBoundingClientRect();
        var offsetLeft = targetPosition.left - (parentPosition.left - parent.scrollLeft);
        var offsetTop = targetPosition.top - (parentPosition.top - parent.scrollTop);
        x = offsetLeft + (targetWidth * leftScalar) - parent.clientWidth * leftScalar;
        y = offsetTop + (targetHeight * topScalar) - parent.clientHeight * topScalar;
        x -= leftOffset;
        y -= topOffset;
        x = Math.max(Math.min(x, parent.scrollWidth - parent.clientWidth), 0);
        y = Math.max(Math.min(y, parent.scrollHeight - parent.clientHeight), 0);
        differenceX = x - parent.scrollLeft;
        differenceY = y - parent.scrollTop;
    }

    return {
        x: x,
        y: y,
        differenceX: differenceX,
        differenceY: differenceY
    };
}

function animate(parent){
    var scrollSettings = parent._scrollSettings;

    if(!scrollSettings){
        return;
    }

    var maxSynchronousAlignments = scrollSettings.maxSynchronousAlignments;

    var location = getTargetScrollLocation(scrollSettings, parent),
        time = Date.now() - scrollSettings.startTime,
        timeValue = Math.min(1 / scrollSettings.time * time, 1);

    if(scrollSettings.endIterations >= maxSynchronousAlignments){
        setElementScroll(parent, location.x, location.y);
        parent._scrollSettings = null;
        return scrollSettings.end(COMPLETE);
    }

    var easeValue = 1 - scrollSettings.ease(timeValue);

    setElementScroll(parent,
        location.x - location.differenceX * easeValue,
        location.y - location.differenceY * easeValue
    );

    if(time >= scrollSettings.time){
        scrollSettings.endIterations++;
        return animate(parent);
    }

    raf(animate.bind(null, parent));
}

function defaultIsWindow(target){
    return target.self === target
}

function transitionScrollTo(target, parent, settings, callback){
    var idle = !parent._scrollSettings,
        lastSettings = parent._scrollSettings,
        now = Date.now(),
        cancelHandler,
        passiveOptions = { passive: true };

    if(lastSettings){
        lastSettings.end(CANCELED);
    }

    function end(endType){
        parent._scrollSettings = null;
        if(parent.parentElement && parent.parentElement._scrollSettings){
            parent.parentElement._scrollSettings.end(endType);
        }

        if(settings.debug){
            console.log('Scrolling ended with type', endType, 'for', parent)
        }

        callback(endType);
        if(cancelHandler){
            parent.removeEventListener('touchstart', cancelHandler, passiveOptions);
            parent.removeEventListener('wheel', cancelHandler, passiveOptions);
        }
    }

    var maxSynchronousAlignments = settings.maxSynchronousAlignments;

    if(maxSynchronousAlignments == null){
        maxSynchronousAlignments = 3;
    }

    parent._scrollSettings = {
        startTime: lastSettings ? lastSettings.startTime : Date.now(),
        endIterations: 0,
        target: target,
        time: settings.time + (lastSettings ? now - lastSettings.startTime : 0),
        ease: settings.ease,
        align: settings.align,
        isWindow: settings.isWindow || defaultIsWindow,
        maxSynchronousAlignments: maxSynchronousAlignments,
        end: end
    };

    if(!('cancellable' in settings) || settings.cancellable){
        cancelHandler = end.bind(null, CANCELED);
        parent.addEventListener('touchstart', cancelHandler, passiveOptions);
        parent.addEventListener('wheel', cancelHandler, passiveOptions);
    }

    if(idle){
        animate(parent);
    }
}

function defaultIsScrollable(element){
    return (
        'pageXOffset' in element ||
        (
            element.scrollHeight !== element.clientHeight ||
            element.scrollWidth !== element.clientWidth
        ) &&
        getComputedStyle(element).overflow !== 'hidden'
    );
}

function defaultValidTarget(){
    return true;
}

module.exports = function(target, settings, callback){
    if(!target){
        return;
    }

    if(typeof settings === 'function'){
        callback = settings;
        settings = null;
    }

    if(!settings){
        settings = {};
    }

    settings.time = isNaN(settings.time) ? 1000 : settings.time;
    settings.ease = settings.ease || function(v){return 1 - Math.pow(1 - v, v / 2);};

    var parent = target.parentElement,
        parents = 1;

    function done(endType){
        parents--;
        if(!parents){
            callback && callback(endType);
        }
    }

    var validTarget = settings.validTarget || defaultValidTarget;
    var isScrollable = settings.isScrollable;

    if(settings.debug){
        console.log('About to scroll to', target)

        if(!parent){
            console.error('Target did not have a parent, is it mounted in the DOM?')
        }
    }

    while(parent){
        if(parent.tagName === 'BODY'){
            parent = parent.ownerDocument;
            parent = parent.defaultView || parent.ownerWindow;
        }

        if(settings.debug){
            console.log('Scrolling parent node', parent)
        }

        if(validTarget(parent, parents) && (isScrollable ? isScrollable(parent, defaultIsScrollable) : defaultIsScrollable(parent))){
            parents++;
            transitionScrollTo(target, parent, settings, done);
        }

        parent = parent.parentElement;

        if(!parent){
            done(COMPLETE)
            break;
        }
    }
};

},{}]},{},[1]);
